core/src/main/java/org/apache/iceberg/deletes/SortingPositionOnlyDeleteWriter.java

```diff
@@ -123,22 +141,42 @@ private DeleteWriteResult writeDeletes(Collection<CharSequence> paths) throws IO
     }
 
     FileWriter<PositionDelete<T>, DeleteWriteResult> writer = writers.get();
+    List<DeleteFile> rewrittenDeleteFile = Lists.newArrayList();
 
     try {
       PositionDelete<T> positionDelete = PositionDelete.create();
       for (CharSequence path : sort(paths)) {
-        // the iterator provides values in ascending sorted order
-        PeekableLongIterator positions = positionsByPath.get(path).getLongIterator();
-        while (positions.hasNext()) {
-          long position = positions.next();
-          writer.write(positionDelete.set(path, position, null /* no row */));
+        PositionDeleteIndex positions = positionsByPath.get(path);
+        PositionDeleteIndex previousPositions = loadPreviousDeletes(path);
+        if (previousPositions != null && previousPositions.isNotEmpty()) {
+          validatePreviousDeletes(previousPositions);
+          positions.merge(previousPositions);
+          rewrittenDeleteFile.addAll(previousPositions.deleteFiles());
         }
+        positions.forEach(position -> writer.write(positionDelete.set(path, position)));
       }
     } finally {
       writer.close();
     }
 
-    return writer.result();
+    DeleteWriteResult writerResult = writer.result();
+    List<DeleteFile> deleteFiles = writerResult.deleteFiles();
+    CharSequenceSet referencedDataFiles = writerResult.referencedDataFiles();
+    return new DeleteWriteResult(deleteFiles, referencedDataFiles, rewrittenDeleteFile);
+  }
+
+  private void validatePreviousDeletes(PositionDeleteIndex index) {
+    Preconditions.checkArgument(
+        index.deleteFiles().stream().allMatch(this::isFileScoped),
+        "Previous deletes must be file-scoped");
+  }
+
+  private boolean isFileScoped(DeleteFile deleteFile) {
+    return ContentFileUtil.referencedDataFile(deleteFile) != null;
+  }
+
+  private PositionDeleteIndex loadPreviousDeletes(CharSequence path) {
+    return loadPreviousDeletes != null ? loadPreviousDeletes.apply(path) : null;
```

Comment:
Nit: Would it make sense to default `loadPreviousDeletes` to be a function implementation which just returns null (I think it'd be a one line lambda `charSequence -> null`? Then I think we could remove this helper method and directly use `loadPreviousDeletes.apply(path)` on line 150.

Reply:
I like that, let me update.